### ISSUE TITLE
[Feature]: Roll back to previous packages version

### ISSUE THREAD

**Issue by jumper047** [progfolio/elpaca#24]
### Feature Description

As far as I know, Straight.el has command `straight-freeze-version`. Is similar feature planned for Elpaca? Ideally, I would love to have ability to roll back packages versions to ones before last upgrade

### Confirmation

- [X] The feature I'm proposing does not already exist in Elpaca

**Comment by progfolio** [comment]
> Is similar feature planned for Elpaca?

Yes. The `elpaca-write-lockfile` and `elpaca-load-lockfile` commands are stubbed out in elpaca.el.
I'm still considering some aspects of the design.

> I would love to have ability to roll back packages versions to ones before last upgrade

My plan is to make general commands that would make this easy to do.
I don't know if I'll provide a "roll back to the the versions just before an update" command specifically, but it will be easy to write one.

**Comment by noctuid** [comment]
It would be nice if these things would be supported:

- A command to write the current hash for a single package
- A key to manually ignore a dirty work tree when processing packages and just write the hash
- A setting to automatically ignore a dirty worktree for certain packages (my main reason for wanting this is that the first Emacs packages I wrote have a lot of unused files I need to clean up but haven't had the time to; I also have todo files in some repos I'd like to clean and add at some point rather than .gitignore)
- An option to to print using lv or something else (so messages/prompts persist even if something else prints to the echo area)
- Some command to validate a configuration/lockfile.  Sometimes a recipe (normally a custom one) is no longer valid due to the repo moving, for example.  I've even seen a couple cases where a hash was no longer valid due to force pushes to master... It may be overkill/unreasonable for elpaca to do this.  You can validate manually by running Emacs with a different home directory, for example.

**Comment by progfolio** [comment]
Rudimentary support for lock files added.

> - A command to write the current hash for a single package

Should be possible via `elpaca-lock-file-functions`.

> - A key to manually ignore a dirty work tree when processing packages and just
>   write the hash

Currently no worktrees are checked. Current hash is written.

> - Some command to validate a configuration/lockfile. Sometimes a recipe
>   (normally a custom one) is no longer valid due to the repo moving, for
>   example. I've even seen a couple cases where a hash was no longer valid due to
>   force pushes to master... It may be overkill/unreasonable for elpaca to do
>   this. You can validate manually by running Emacs with a different home
>   directory, for example.

I think attempting a bootstrap with the `--init-directory` option or using the `elpaca-test` macro should work for this.

**Comment by progfolio** [comment]
> I would love to have ability to roll back packages versions to ones before last upgrade

On latest master you should be able to:

1. `M-x elpaca-write-lock-file` to the desired path.
2. Update desired packages.
3. `M-x elpaca-delete` any packages you want to roll back. 
4. `(setq elpaca-lock-file "/path/to/file/from/step/one.eld")`
5. `M-x restart-emacs`

At that point the packages should be installed in the state they were in at step 1.
This workflow could be automated to some extent by interested parties.
I have plans to do more work on lock files in the future, but it's enough to get started with.
